src/neo4j_genai/neo4j_queries.py

```diff
@@ -12,51 +12,169 @@
     retrieval_query: Optional[str] = None,
-):
-    query_map = {
-        SearchType.VECTOR: "".join(
-            [
-                "CALL db.index.vector.queryNodes($index_name, $top_k, $query_vector) ",
-                "YIELD node, score ",
-                get_query_tail(retrieval_query, return_properties),
-            ]
-        ),
-        SearchType.HYBRID: "".join(
-            [
-                "CALL { ",
-                "CALL db.index.vector.queryNodes($vector_index_name, $top_k, $query_vector) ",
-                "YIELD node, score ",
-                "RETURN node, score UNION ",
-                "CALL db.index.fulltext.queryNodes($fulltext_index_name, $query_text, {limit: $top_k}) ",
-                "YIELD node, score ",
-                "WITH collect({node:node, score:score}) AS nodes, max(score) AS max ",
-                "UNWIND nodes AS n ",
-                "RETURN n.node AS node, (n.score / max) AS score ",
-                "} ",
-                "WITH node, max(score) AS score ORDER BY score DESC LIMIT $top_k ",
-                get_query_tail(
-                    retrieval_query, return_properties, "RETURN node, score"
-                ),
-            ]
-        ),
-    }
-    return query_map[search_type]
-
-
-def get_query_tail(
+    node_label: Optional[str] = None,
+    embedding_node_property: Optional[str] = None,
+    embedding_dimension: Optional[int] = None,
+    filters: Optional[dict[str, Any]] = None,
+) -> tuple[str, dict[str, Any]]:
+    """Build the search query, including pre-filtering if needed, and return clause.
+
+    Args
+        search_type: Search type we want to search for:
+        return_properties (list[str]): list of property names to return.
+            It can't be provided together with retrieval_query.
+        retrieval_query (str): the query to use to retrieve the search results
+            It can't be provided together with return_properties.
+        node_label (str): node label we want to search for
+        embedding_node_property (str): the name of the property holding the embeddings
+        embedding_dimension (int): the dimension of the embeddings
+        filters (dict[str, Any]): filters used to pre-filter the nodes before vector search
+
+    Returns:
+        tuple[str, dict[str, Any]]: query and parameters
+
+    """
+    if search_type == SearchType.HYBRID:
+        if filters:
+            raise Exception("Filters is not supported with Hybrid Search")
+        query = _get_hybrid_query()
+        params = {}
+    elif search_type == SearchType.VECTOR:
+        query, params = _get_vector_query(
+            filters, node_label, embedding_node_property, embedding_dimension
+        )
+    else:
+        raise ValueError(f"Search type is not supported: {search_type}")
+    query_tail = _get_query_tail(
+        retrieval_query, return_properties, fallback_return="RETURN node, score"
+    )
+    return " ".join([query, query_tail]), params
```

Comment:
I think f"{query} {query_tail}" is more readable here

Reply:
I don't agree, I think it's better as it is :)

Reply:
I'd personally do `" ".join([a, b])` if we already have a list to begin with and pass it in, which is why I'd prefer `f"{query} {query_tail}"`. 

However, this is subjective. I'm happy to go either way as long as we keep preferences consistent across the codebase

Reply:
I do not have a strong opinion on this one, I kept the join that was there in the previous version. Now, f-string seems to be slightly faster, and I do not think we plan to add more items to this list (post-filtering can be performed in the retrieval query), so we can move to f-string.

Reply:
Actually, updated to f-string for consistency with other parts of the code (sorry Oskar :) )